apple/swift-docc
# drop an "Extended Symbol" page when its children are curated elsewhere

Bug/issue #, if applicable: rdar://107729630

## Summary

When Swift-DocC gets symbol graphs with extension block symbols, it creates "Extended Module/Class/Structure/etc" symbols to automatically curate the extension symbols into. However, if the extension symbols are then curated in the rest of the documentation hierarchy, these extra symbols are left in the topic graph and create empty pages that don't need to be there. This PR adds a pass to `DocumentationContext.register(_:)` to drop these symbols if all their children have been curated elsewhere. (This pass is done recursively, so "Extended Module" pages will drop if all their child "Extended Class/Structure/etc" pages are also dropped.)

## Dependencies

None

## Testing

The following code was used to test this functionality:

```swift
/// a namespace for organization.
///
/// ## Topics
///
/// - ``Swift/Array/asdf``
public enum MyNamespace {}

public extension Array {
  var asdf: Int { count }
}
```

Steps:
1. Build a symbol graph for the above code that includes extension block symbols. (The code above was used to create the `ModuleWithSingleExtension` test fixture added in this PR, which can be used as a shortcut.)
2. Add the resulting symbol graphs to a documentation catalog and generate documentation.
3. Ensure that the resulting documentation doesn't contain the "Extended Module" page for `Swift`, nor the "Extended Structure" page for `Array`.

## Checklist

Make sure you check off the following items. If they cannot be completed, provide a reason.

- [ n/a ] Updated documentation if necessary
